### PR TITLE
ARROW-1701: [Serialization] Support zero copy PyTorch Tensor serialization

### PR DESCRIPTION
**Pull request by pcmoritz**
This also restructures the code such that it is easier to reset the default serialization context back to the initial state after more handlers have been registered.

### CODE REVIEW

File: python/pyarrow/tests/test_serialization.py [hunk 1]
```diff
@@ -285,6 +285,16 @@ def test_datetime_serialization(large_memory_map):
...
+def test_torch_serialization(large_memory_map):
+    pytest.importorskip("torch")
+    import torch
+    with pa.memory_map(large_memory_map, mode="r+") as mmap:
+        # These are the only types that are supported for the
+        # PyTorch to NumPy conversion
+        for t in ["float32", "float64",
+                  "uint8", "int16", "int32", "int64"]:
+            obj = torch.from_numpy(np.random.randn(1000).astype(t))
+            serialization_roundtrip(obj, mmap)
```
Comment (robertnishihara) [inline python/pyarrow/tests/test_serialization.py]: I wouldn't expect our `serialization_roundtrip` function to work on this. You may need to either modify the `assert_equal` function to handle torch tensors (don't forget to check both the values and the dtypes), or you may need to just explicitly compare the dtypes and the values.

**Comment by robertnishihara** [comment]
This looks good to me, except I'm not sure the test is correct.

File: python/pyarrow/tests/test_serialization.py (hunk 1 above)
Comment (pcmoritz) [reply python/pyarrow/tests/test_serialization.py]: I fixed the tests (and made sure torch.equals only returns true if the datatypes also agree)

**Comment by pcmoritz** [comment]
+1, this can be merged

**Comment by wesm** [comment]
Can you instead use "[Python]" for the category in the PR title?

**Comment by wesm** [comment]
+1, merging (not big deal about my PR title nitpick, but good to keep the changelog clean)

**Comment by ppwwyyxx** [comment]
Could you make pyarrow NOT do `import torch` by default?
Pytorch is a giant library with multiple C extensions, so importing it has a lot of side-effect especially about symbol version conflict, e.g. issue 3059, 926, 4111 in pytorch. 

I've seen other conflicting issues created by `import torch` in addition to those above. More recently I found this import came from pyarrow after debugging my segfaults for a long time.

**Comment by robertnishihara** [comment]
@ppwwyyxx good point, I think we should get rid of the global variables `_default_serialization_context` and `pandas_serialization_context` and replace them with functions `default_serialization_context()` and `pandas_serialization_context()`.

This will also make it faster to do `import pyarrow`.

**Comment by wesm** [comment]
Seems fine to me. Can someone open a JIRA about this?

**Comment by robertnishihara** [comment]
https://issues.apache.org/jira/browse/ARROW-2024

**Comment by ppwwyyxx** [comment]
Can we push the update to pypi?

**Comment by wesm** [comment]
We're about to start the 0.9.0 release vote. So updated packages ought to be up next week sometime

**Pull request closed without merging**